File: Java Projects v2/Project 2_1.java
import java.util.Scanner;

public class PalindromeChecker {
    // Function to check palindrome iteratively
    public static boolean isPalindromeIterative(String str) {
        int left = 0, right = str.length() - 1;
        while (left < right) {
            if (str.charAt(left) != str.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
    
    // Function to check palindrome recursively
    public static boolean isPalindromeRecursive(String str, int left, int right) {
        if (left >= right) return true;
        if (str.charAt(left) != str.charAt(right)) return false;
        return isPalindromeRecursive(str, left + 1, right - 1);
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter a string or number to check for palindrome: ");
        String input = scanner.nextLine();
        scanner.close();
        
        // Removing spaces and converting to lowercase for case-insensitive check
        String processedInput = input.replaceAll("\\s+", "").toLowerCase();
        
        System.out.println("\nUsing Iterative Approach:");
        if (isPalindromeIterative(processedInput))
            System.out.println("\"" + input + "\" is a palindrome.");
        else
            System.out.println("\"" + input + "\" is not a palindrome.");
        
        System.out.println("\nUsing Recursive Approach:");
        if (isPalindromeRecursive(processedInput, 0, processedInput.length() - 1))
            System.out.println("\"" + input + "\" is a palindrome.");
        else
            System.out.println("\"" + input + "\" is not a palindrome.");
    }
}

File: Cpp Projects v2/Project 2_1.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Function to check if a number is prime iteratively
bool isPrimeIterative(int n) {
    if (n < 2) return false;
    for (int i = 2; i <= sqrt(n); i++) {
        if (n % i == 0) return false;
    }
    return true;
}

// Function to check if a number is prime recursively
bool isPrimeRecursive(int n, int i = 2) {
    if (n < 2) return false;
    if (i > sqrt(n)) return true;
    if (n % i == 0) return false;
    return isPrimeRecursive(n, i + 1);
}

int main() {
    int num;
    cout << "Enter a number to check if it's prime: ";
    cin >> num;

    cout << "\nUsing Iterative Approach:" << endl;
    if (isPrimeIterative(num))
        cout << num << " is a prime number." << endl;
    else
        cout << num << " is not a prime number." << endl;

    cout << "\nUsing Recursive Approach:" << endl;
    if (isPrimeRecursive(num))
        cout << num << " is a prime number." << endl;
    else
        cout << num << " is not a prime number." << endl;

    return 0;
}